app/src/main/java/com/jalasoft/routesapp/util/algorithm/RouteCalculator.kt

```diff
@@ -0,0 +1,116 @@
+                        val route = Route(line.name, newLine, line.start, newStops)
+                        availableTransport.add(AvailableTransport(null, mutableListOf(TransportWithLine(transport.name, route))))
+                    }
+                } else {
+                    if (nearestStopToDestination.distanceTo(destination).toDouble() <= minDistance) {
+                        val intersectionStopIndex = line.stops.indexOfFirst {
+                            it.latitude == nearestStopToDestination.latitude && it.longitude == nearestStopToDestination.longitude
+                        }
+                        val intersectionRoutePointIndex = line.routePoints.indexOfFirst {
+                            it.latitude == nearestStopToDestination.latitude && it.longitude == nearestStopToDestination.longitude
+                        }
+
+                        val newStops = line.stops.slice(0..intersectionStopIndex)
+                        val newRoutePoints = line.routePoints.slice(0..intersectionRoutePointIndex)
+                        val subLine = Route(line.name, newRoutePoints, line.start, newStops)
+                        candidates.destinationList.add(TransportWithLine(transport.name, subLine))
+                    }
+
+                    if (nearestStopToOrigin.distanceTo(origin).toDouble() <= minDistance) {
+                        val intersectionStopIndex = line.stops.indexOfFirst {
+                            it.latitude == nearestStopToOrigin.latitude && it.longitude == nearestStopToOrigin.longitude
+                        }
+                        val intersectionRoutePointIndex = line.routePoints.indexOfFirst {
+                            it.latitude == nearestStopToOrigin.latitude && it.longitude == nearestStopToOrigin.longitude
+                        }
+
+                        val newStops = line.stops.slice(intersectionStopIndex until line.stops.size)
+                        val newRoutePoints = line.routePoints.slice(intersectionRoutePointIndex until line.routePoints.size)
+                        val subline = Route(line.name, newRoutePoints, line.start, newStops)
+
+                        candidates.originList.add(TransportWithLine(transport.name, subline))
+                    }
+                }
+            }
+        }
+
+        for (routeFromOrigin in candidates.originList) {
+            for (routeFromDestiny in candidates.destinationList) {
+                for (stop in routeFromDestiny.line.stops) {
+                    val nearestStop = routeFromOrigin.line.stops.minWith(Comparator.comparingDouble { it.distanceTo(stop).toDouble() })
+                    val indexOfNearestStop = routeFromOrigin.line.stops.indexOf(nearestStop)
+                    if (nearestStop.distanceTo(stop).toDouble() <= minDistanceBtwStops) {
+                        // Line A
+                        val indexOfRoutesA = routeFromOrigin.line.routePoints.indexOfFirst {
```

Comment:
you can also here make a method to obtain lineA and another one to obtain lineB.
All of this to reduce cognitive complexity for the algorithm

Reply:
Solved